File: Benchmarks/Core/Components/InputComponent.cs
namespace Benchmarks.Core.Components
{
    [Flags]
    public enum InputInUse
    {
        None = 0,
        HasMouse = 1 << 0,
        HasGamepad = 1 << 1,
        HasKeyboard = 1 << 2,
    }

    public enum ActionEntityInput
    {
        None,
        MoveUp,
        MoveDown,
        MoveLeft,
        MoveRight,
        Jump,
        Attack,
        Interact,
    }

    public struct InputComponent()
    {
        private int InputsInUse { get; set; } = 0;

        public Dictionary<ActionEntityInput, int> KeyBoard { get; set; }
        public Dictionary<ActionEntityInput, int> GamePad { get; set; }
        public Dictionary<ActionEntityInput, int> Mouse { get; set; }
        public int GamePadIndex { get; set; }

        public bool HasInput(InputInUse input) => (InputsInUse & (int)input) != 0;
        public void AddInput(InputInUse input) => InputsInUse |= (int)input;
        public void RemoveInput(InputInUse input) => InputsInUse &= ~(int)input;
    }
}

File: Benchmarks_FSharp/Core/Components/InputComponent.fs
namespace Benchmarks.Core.Components

open System
open System.Collections.Generic

[<Flags>]
type InputInUse =
    | None = 0
    | HasMouse = 1
    | HasGamepad = 2
    | HasKeyboard = 4

type ActionEntityInput =
    | None = 0
    | MoveUp = 1
    | MoveDown = 2
    | MoveLeft = 3
    | MoveRight = 4
    | Jump = 5
    | Attack = 6
    | Interact = 7

[<Struct>]
type InputComponent =
    val mutable private InputsInUse: int
    val mutable KeyBoard: Dictionary<ActionEntityInput, int>
    val mutable GamePad: Dictionary<ActionEntityInput, int>
    val mutable Mouse: Dictionary<ActionEntityInput, int>
    val mutable GamePadIndex: int

    member this.HasInput(input: InputInUse) = (this.InputsInUse &&& int input) <> 0
    member this.AddInput(input: InputInUse) = this.InputsInUse <- this.InputsInUse ||| int input
    member this.RemoveInput(input: InputInUse) = this.InputsInUse <- this.InputsInUse &&& ~~~(int input)
